crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2045,52 @@ impl<'db> Class<'db> {
         self.iter_mro(db).contains(&ClassBase::Class(other))
     }
 
+    /// Return the metaclass of this class.
+    pub(crate) fn metaclass(self, db: &'db dyn Db) -> Result<Type<'db>, MetaclassError> {
+        let mut base_classes = self.explicit_bases(db).filter_map(|base| match base {
+            Type::ClassLiteral(class) => Some(class),
+            _ => None,
+        });
+
+        // Identify the metaclasses of all the class's bases, plus the class's explicit metaclass.
+        let metaclass = if let Some(metaclass) = self.explicit_metaclass(db) {
+            metaclass
+        } else if let Some(base_class) = base_classes.next() {
+            base_class.metaclass(db)?
+        } else {
+            KnownClass::Type.to_instance(db)
+        };
+
+        let Type::ClassLiteral(mut candidate) = metaclass else {
+            // If the metaclass is not a class, return it directly.
+            return Ok(metaclass);
+        };
+
+        // Otherwise, reconcile the metaclasses of the class's bases.
+        //
+        // See: https://github.com/python/cpython/blob/83ba8c2bba834c0b92de669cac16fcda17485e0e/Objects/typeobject.c#L3629-L3663
+        for base_class in base_classes {
+            let Type::ClassLiteral(metaclass) = base_class.metaclass(db)? else {
+                continue;
+            };
```

Comment:
Not totally certain what to do here.

Reply:
I think this is the right thing to do; it matches what happens at runtime. It would be good to have a test for this behavior, too.

Reply:
I'm not sure I agree with this. If a non-`type`-subclass is used as the metaclass, I think it's more likely than not that `__class__` will be something completely different to the object passed as the `metaclass=` keyword:

```pycon
>>> class Foo(metaclass=lambda *args: 42): ...
... 
>>> Foo
42
>>> Foo.__class__
<class 'int'>
```

I think with the current logic we'd infer `Foo.__class__` here as being `<lambda function>` or something? But that's just incorrect.

Even if the metaclass is something that returns a class (I think the more common case, really!), it's almost certainly going to be a callable object, and the `__class__` of the created class is not going to be the same as the callable object itself.

I think honestly we don't have a clue what's going on if a non-`type`-subclass is passed as the `metaclass=` keyword, so emitting a diagnostic and inferring `Unknown` as the value of `__class__` might be better.

Reply:
Oh! Great point.

I think this is something we should address in the future once we have call signature checking, and what we should do here is a) verify that the metaclass is something callable which accepts the type-new arguments, and then b) return the return type of that callable.

For now returning `Type::Todo` with a todo comment to that effect is probably best.